org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/SaveActionExtensions.scala

```diff
@@ -153,31 +147,63 @@ trait SaveActionExtensions extends HasLogger {
   }
 
   /**
-   * Applies all save actions that extends [[DocumentSupport]].
+   * Applies all save actions that extend [[DocumentSupport]].
    */
   private def applyDocumentExtensions(udoc: IDocument): Unit = {
     for ((setting, ext) <- documentSaveActions if isEnabled(setting.id)) {
       val doc = new TextDocument(udoc)
       val instance = ext(doc)
-      performExtension(instance, udoc) {
+      performDocumentExtension(instance, udoc) {
         instance.perform()
       }
     }
   }
 
   /**
-   * Applies all save actions that extends [[CompilerSupport]].
+   * Applies all save actions that extend [[CompilerSupport]].
    */
   private def applyCompilerExtensions(udoc: IDocument): Unit = {
-    for ((setting, ext) <- compilerSaveActions if isEnabled(setting.id)) {
-      createExtensionWithCompilerSupport(ext) foreach { instance =>
-        performExtension(instance, udoc) {
-          instance.global.asInstanceOf[IScalaPresentationCompiler].asyncExec {
-            instance.perform()
-          }.getOrElse(Seq())()
+    val timeout = saveActionTimeout
+
+    def loop(xs: Seq[(SaveActionSetting, CompilerSupportCreator)]): Unit = xs match {
+      case Nil ⇒
+
+      case (setting, ext) :: xs if isEnabled(setting.id) ⇒
+        val res = FutureUtils.performWithTimeout(timeout) {
+          EclipseUtils.withSafeRunner(s"An error occurred while executing compiler driven save action.") {
+            createExtensionWithCompilerSupport(ext) map { instance =>
+              instance.global.asInstanceOf[IScalaPresentationCompiler].asyncExec {
+                instance.perform()
+              }.getOrElse(Seq())()
+            }
+          }.flatten.getOrElse(Seq())
         }
-      }
+
+        res match {
+          case Success(changes) ⇒
+            EclipseUtils.withSafeRunner(s"An error occurred while applying changes of save action '${setting.id}'.") {
+              applyChanges(setting.id, changes, udoc)
+            }
+            loop(xs)
+
+          case Failure(_) ⇒
+            eclipseLog.error(s"""|
+               |A save action that relies on compiler support didn't complete in a
+               | given time, it had $timeout time to complete. Because it is
+               | unlikely that further save actions that rely on compiler support will
+               | complete in time, no further save actions are executed. Please consider
+               | to disable save actions that rely on the compiler. The performed save
+               | action itself couldn't be aborted, therefore if you know that it may
+               | never complete in future, you may wish to restart your Eclipse to
+               | clean up your VM.
+               |
+               |""".stripMargin.replaceAll("\n", ""))
+        }
+
+      case _ :: xs ⇒
```

Comment:
Same as above, `::` matches only lists. Also, `xs` shadows `xs` method argument, maybe rename it to `tail` (same above)?